### PR TITLE
Keep track of all parents in the AST map

### PR DESCRIPTION
**Pull request by nrc**
r? @Manishearth

### CODE REVIEW

File: src/librustc/ast_map/mod.rs [hunk 1]
```diff
@@ -121,14 +121,14 @@ pub enum Node<'ast> {
...
-/// Represents an entry and its parent Node ID
+/// Represents an entry and its parent NodeID.
 /// The odd layout is to bring down the total size.
 #[derive(Copy, Debug)]
 enum MapEntry<'ast> {
     /// Placeholder for holes in the map.
     NotPresent,
 
-    /// All the node types, with a parent ID.
+    /// All the node types, with a parent and scope ID.
```
Comment (eddyb) [inline src/librustc/ast_map/mod.rs]: This isn't true anymore AFAICT.

File: src/librustc/ast_map/mod.rs [hunk 2]
```diff
@@ -283,10 +283,84 @@ impl<'ast> Map<'ast> {
...
+    /// never appear as the parent_node. So you can always walk the parent_nodes
+    /// from a node to the root of the ast (unless you get the same id back here
+    /// that can happen if the id is not in the map itself or is just weird).
+    pub fn get_parent_node(&self, id: NodeId) -> NodeId {
+        self.find_entry(id).and_then(|x| x.parent_node()).unwrap_or(id)
+    }
+
+    /// If there is some error when walking the parents (e.g., a node does not
+    /// have a parent in the map or a node can't be found), then we return the
+    /// last good node id we found.
```
Comment (Manishearth) [inline src/librustc/ast_map/mod.rs]: Also note that reaching the root is considered a success (perhaps we should make it an error instead?)

File: src/librustc/ast_map/mod.rs [hunk 3]
```diff
@@ -283,10 +283,84 @@ impl<'ast> Map<'ast> {
...
+                None => {
+                    return Err(parent_node);
+                }
+            }
+            id = parent_node;
+        }
+    }
+
     /// Retrieve the parent NodeId for `id`, or `id` itself if no
     /// parent is registered in this map.
```
Comment (Manishearth) [inline src/librustc/ast_map/mod.rs]: nit: Cleared documentation on what this fetches

File: src/librustc/ast_map/mod.rs [hunk 4]
```diff
@@ -590,15 +664,15 @@ impl<'a, 'ast> Iterator for NodesMatchingSuffix<'a, 'ast> {
...
             }
             self.idx += 1;
-            let (p, name) = match self.map.find_entry(idx) {
-                Some(EntryItem(p, n))       => (p, n.name()),
-                Some(EntryForeignItem(p, n))=> (p, n.name()),
-                Some(EntryTraitItem(p, n))  => (p, n.name()),
-                Some(EntryImplItem(p, n))   => (p, n.name()),
-                Some(EntryVariant(p, n))    => (p, n.name()),
+            let name = match self.map.find_entry(idx) {
+                Some(EntryItem(_, n))       => n.name(),
```
Comment (Manishearth) [inline src/librustc/ast_map/mod.rs]: Aha, this is where I messed up in my own attempt.

**Comment by Manishearth** [comment]
LGTM modulo nits

File: src/librustc/ast_map/mod.rs (hunk 2 above)
Comment (nrc) [inline src/librustc/ast_map/mod.rs]: That is success in most cases, it means we started looking up from an item under the crate root

Comment (Manishearth) [inline src/librustc/ast_map/mod.rs]: Sure, but that should be explicit in the doc comment :)

**Pull request merged**